connectors/kafka/src/main/java/forklift/connectors/KafkaController.java

```diff
@@ -0,0 +1,225 @@
+     * @param record the record to acknowledge
+     * @return true if the record should be processed, else false
+     * @throws InterruptedException
+     * @throws JMSException
+     */
+    public boolean acknowledge(ConsumerRecord<?, ?> record) throws InterruptedException, JMSException {
+        log.debug("Acknowledge message with topic {} partition {} offset {}", record.topic(), record.partition(), record.offset());
+        return running && this.acknowlegmentHandler.acknowledgeRecord(record);
+    }
+
+    /**
+     * Starts the controller.  Must be called before any other methods.
+     * <p>
+     * <strong>WARNING: </strong>Avoid starting a service which has already run and been stopped.  No attempt is made
+     * to recover a stopped controller.
+     */
+    public void start() {
+        running = true;
+        executor.submit(() -> controlLoop());
+    }
+
+    /**
+     * Stops the controller.
+     *
+     * @param timeout  the time to wait for the service to stop
+     * @param timeUnit the TimeUnit of timeout
+     * @throws InterruptedException
+     */
+    public void stop(long timeout, TimeUnit timeUnit) throws InterruptedException {
+        running = false;
+        kafkaConsumer.wakeup();
+        executor.shutdownNow();
+        executor.awaitTermination(timeout, timeUnit);
+    }
+
+    private void controlLoop() {
+        try {
+            while (running) {
+                boolean updatedAssignment = false;
+                if (topics.size() == 0) {
+                    //check if the last remaining topic was removed
+                    if (kafkaConsumer.assignment().size() > 0) {
+                        kafkaConsumer.unsubscribe();
+                    }
+                    synchronized (topicsMonitor) {
+                        //recheck wait condition inside synchronized block
+                        if (topics.size() == 0) {
+                            //pause the polling thread until a topic comes in
+                            topicsMonitor.wait();
+                        }
+                    }
+                }
+                if (topicsChanged) {
+                    topicsChanged = false;
+                    kafkaConsumer.subscribe(topics, new RebalanceListener());
+                    updatedAssignment = true;
+                }
+                ConsumerRecords<?, ?> records = kafkaConsumer.poll(100);
+                if (updatedAssignment) {
+                    this.acknowlegmentHandler.addPartitions(kafkaConsumer.assignment());
+                }
+                if (records.count() > 0) {
+                    log.debug("Adding: " + records.count() + " to record stream");
+                    messageStream.addRecords(consumerRecordsToKafkaMessages(records));
+                }
+                Map<TopicPartition, OffsetAndMetadata> offsetData = this.acknowlegmentHandler.flushAcknowledged();
+                if (offsetData.size() > 0) {
+                    String offsetDescription =
```

Comment:
I would rather not do this string construction all the time if it only matters for debug logs.

It would be cool if we could just throw `lazy` before this and call it good, but alas.